packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -507,71 +492,121 @@ rocksdb::Status ArbCore::reorgCheckpoints(
+                        std::cerr
+                            << "Last L2 block saved to database (no machine): "
+                            << machine_output.l2_block_number << std::endl;
+                    }
+                    selected_machine_output = machine_output;
+                }
                 if (check_output(machine_output)) {
-                    if (!selected_machine_output.has_value()) {
-                        // Save selected output to know how much machine needs
-                        // to be executed if it behind
-                        selected_machine_output = machine_output;
+                    // All outdated checkpoints have been removed
+                    // TODO: Do not check cache every checkpoint without machine
+                    auto mach = combined_sideload_cache.atOrBeforeGas(
+                        machine_output.arb_gas_used, 0, 0,
+                        coreConfig.checkpoint_load_gas_cost,
+                        coreConfig.checkpoint_max_execution_gas);
+                    if (mach != nullptr) {
+                        // Found machine in cache
+                        setup = std::make_unique<MachineThread>(
+                            mach->machine_state);
+                        break;
                     }
 
-                    if (std::holds_alternative<std::unique_ptr<MachineThread>>(
-                            setup)) {
-                        // Checkpoint without machine, but machine restored
-                        // from old_machine_cache
-                        break;
+                    if (!selected_machine_output.has_value()) {
+                        // Save first selected output to know how much machine
+                        // needs to be executed if it behind
+                        selected_machine_output = machine_output;
                     }
 
-                    // Need to get machine from older checkpoint without
-                    // full reorg because current checkpoint does not
-                    // have a valid machine
-                    checkpoint_it->Prev();
+                    // Continue iterating until checkpoint with machine is found
                     continue;
                 }
-            }
+            } else {
+                // Checkpoint with machine
+                auto checkpoint =
+                    std::get<MachineStateKeys>(checkpoint_variant);
+                if (initial_start && !selected_machine_output.has_value()) {
+                    // Initial start needs to seed cache through last entry
+                    // in database
+                    if (coreConfig.debug) {
+                        std::cerr << "Last L2 block saved to database: "
+                                  << checkpoint.output.l2_block_number
+                                  << std::endl;
+                    }
+                    selected_machine_output = checkpoint.output;
+                }
```

Comment:
Would be nice to lift this outside the if-else by using a function like https://github.com/OffchainLabs/arbitrum/blob/55b682d908debbfa76a3561a2258656a58844f2d/packages/arb-avm-cpp/data_storage/include/data_storage/executioncursor.hpp#L79 to extract the `MachineOutput` and remove the duplication between the two versions of this code

Reply:
We should be able to lift the `check_output` as well once we have a MachineOutput

Reply:
And maybe also `isValid` which is currently only used on full machine checkpoints but should probably be done in both cases. Currently this could cause issues handling L1 reorgs I think